pkg/scheduler/flavorassigner/flavorassigner.go

```diff
@@ -363,17 +386,37 @@ func (a *FlavorAssigner) assignFlavors(log logr.Logger, counts []int32) Assignme
 			}
 			psAssignment.append(flavors, status)
 		}
-		if features.Enabled(features.TopologyAwareScheduling) {
-			if a.wl.Obj.Spec.PodSets[i].TopologyRequest != nil {
-				assignTopology(log, &psAssignment, a.cq, a.wl.TotalRequests[i], &a.wl.Obj.Spec.PodSets[i])
-			}
-		}
 
 		assignment.append(podSet.Requests, &psAssignment)
 		if psAssignment.Status.IsError() || (len(podSet.Requests) > 0 && len(psAssignment.Flavors) == 0) {
 			return assignment
 		}
 	}
+	if assignment.RepresentativeMode() == NoFit {
+		return assignment
+	}
+
+	if features.Enabled(features.TopologyAwareScheduling) {
+		tasRequests := assignment.TopologyRequestsFor(a.wl, requests, a.cq)
+		if assignment.RepresentativeMode() == Fit {
+			result := a.cq.FindTopologyAssignments(tasRequests)
+			if failure := result.Failure(); failure != nil {
+				// There is at least one PodSet which does not fit
+				psAssignment := assignment.podSetByName(failure.PodSetName)
+				psAssignment.reason(failure.Reason)
+				// clear flavors to make the PodSet assignment as NoFit
+				psAssignment.Flavors = nil
```

Comment:
Should we set `nil` here instead of `NoFit` given that, in `RepresentativeMode()` method we compare different modes?

Reply:
Actually, in this line is for assigning the Flavors. Clearing the flavors as nil will actually result in `NoFit` mode for the PodSetAssignment. The trick to re-compute the assignment is to clear the cached represenativeMode .

There are alternative approaches, such as setting the `representativeMode=NoFit` without clearing the flavors. This would work too, but I'm not sure which one is better. I would propose to revisit this in a follow up PR where we have 2 possible options Preempt or NoFit.